spanish grammar included / en passant bug fixed

diegovalcarce opened the pull request:
Hi, Max. Nice project!
I just added spanish grammar and fixed little bug at "en passant" capture.

* Two new files with spanish grammar text.
* contructor modification to accept new language parameter.
* bug en passant: when a pawn captures in line five, the final rank must be sixth, not fourth. FIXED!

Thank you!

diegovalcarce committed:
spanish grammar included / en passant bug fixed

* Two new files with spanish grammar text
* contructor modification to accept new language parameter
* bug en passant: when a pawn captures in line five, the final rank must be sixth, not fourth. FIXED!

maxwellhaydn commented:
This is fantastic, thank you! I quickly scanned through the changes and they look good, but I'll look more carefully and add any comments if I think anything needs to be fixed before merging. Thank you again!

maxwellhaydn commented on src/grammar/textToSan:
```diff
@@ -82,7 +82,7 @@ en_passant
                   break;
               case '5':           // white capturing black
-                  final_rank = 4;
+                  final_rank = 6;
```
Good catch! Can you please put this en passant fix in a separate pull request? That way I can merge it immediately while reviewing the more extensive changes for supporting multiple languages.

maxwellhaydn commented on src/index.js:
```diff
@@ -136,18 +138,25 @@ const generateRuleText = (rule, aliases) => {
 const ChessNLP = class ChessNLP {
 
-    constructor({ aliases } = { aliases: {} }) {
+    constructor({ aliases = {}, language = 'en' }) {
```
This causes an exception if the constructor is called with no arguments, e.g.

```
import ChessNLP from 'chess-nlp';

const parser = new ChessNLP();
// TypeError: Cannot read property 'aliases' of undefined
```

Can be fixed like this:

```
    constructor({ aliases = {}, language = 'en' } = {}) {
```

maxwellhaydn commented on README.md:
```diff
@@ -145,3 +143,78 @@ isn't a valid chess move.
+
+    const options = {
+        language: 'es'
+        }
```
There's an extraneous curly brace here.

maxwellhaydn commented on README.md:
```diff
@@ -145,3 +143,78 @@ isn't a valid chess move.
+            3: ['iii']
+        },
+        language: 'es'
+        }
```
Another extraneous curly brace.

maxwellhaydn commented on src/index.js:
```diff
@@ -2,34 +2,36 @@ import peg from 'pegjs';
     king: {
         name: 'king',
-        defaultTerms: ['king'],
+        defaultTerms: ['king', 'rey'],
```
Now I see that the way I did the aliases for pieces makes it a little tricky to add multiple languages. This works, but it's not ideal. I'll try to think of a way that doesn't mix the different languages in the same grammar.

maxwellhaydn replied:
I thought about this more and one possibility would be to move the default terms and aliases into a separate config file, e.g. aliases.json:

```
{
    "en": {
        "king": {
            "defaultTerms": ["king"],
            ...
        },
        ...
        "one": {
            "defaultTerms": ["1", "one"]
            ...
        },
        ...
    },
    "es": {
        "king": {
            "defaultTerms": ["rey"],
            ...
        },
        ...
        "one": {
            "defaultTerms": ["1", "uno"]
            ...
        },
        ...
    },
    ...
}
```

It could be done with a single config file, or with a separate config file per language, e.g. with a directory structure like this:

```
+ grammar
    + en
        + sanToText
        + textToSan
        + aliases.json
    + es
        + sanToText
        + textToSan
        + aliases.json
    ...
```

diegovalcarce replied:
Agree! I like separate config file per language solution.

maxwellhaydn commented on README.md:
```diff
@@ -145,3 +143,78 @@ isn't a valid chess move.
+
+    e4      -> e4
+    hxg2    -> h por g2
+    axb8=Q  -> a por b8 igual dama
```
Is "igual" really the most natural way to say this in Spanish? I'm not a native speaker, but maybe "promociona a dama" or "corona con dama" would be a more natural way to describe the move in spoken language. "igual" sounds like you're just reading the algebraic notation out loud. (I see that "promociona" is included in the grammar file.)

maxwellhaydn reviewed (changes requested):
Overall, this is good. The main changes I'd suggest would be:

* Split the en passant fix into a separate PR so I can merge it immediately
* Add tests for the Spanish grammar
* Make it so aliases are language specific, so instead of "rey" and "king" both being accepted in English and Spanish, only "king" is accepted in English and only "rey" in Spanish. The same applies to rank numbers; "one" should only be accepted in English and "uno" should only be accepted in Spanish. This is more an issue with how I implemented aliases, not so much with your change. I'll try to think of a better way to handle this.

See the comments on the particular files for additional minor suggestions.

diegovalcarce committed:
Merge branch 'master' of https://github.com/maxwellhaydn/chess-nlp

diegovalcarce committed:
Spanish grammar included

diegovalcarce commented:
Hi, i did some changes:

- Aliases and grammar in separated files.
- Fixed the constructor (when the constructor was called with no arguments).
- Added "corona" in promotion. It seems you know more spanish word than me. :-)
- Added tests for spanish grammar.

How do you see?
Cheers!!

maxwellhaydn commented:
Diego, this looks really great! I’ll look at it more closely tomorrow and let you know.

maxwellhaydn reviewed (approved):
Looks good, thanks for making those changes!

maxwellhaydn commented:
@all-contributors please add diegovalcarce for code, tests, and docs

maxwellhaydn commented:
Published to NPM in version 1.6.0. Thanks again!

maxwellhaydn commented:
Oops, I also need to add diegovalcarce to credit you for your translation and bug report. cc. @all-contributors
